lib/cc/engine/grep.rb

```diff
@@ -0,0 +1,106 @@
+# frozen_string_literal: true
+
+require "json"
+require "open3"
+
+module CC
+  module Engine
+    class Grep
+      MATCH_START = "\e[01;31m\e[K"
+      MATCH_END = "\e[m\e[K"
+      MARKER_RE = /\e\[.*?\e\[K/
+      MATCH_RE = /#{Regexp.escape MATCH_START}(?<str>.*?)#{Regexp.escape MATCH_END}/.freeze
+
+      def initialize(root:, config: {}, io: $stdout)
+        @root = root
+        @config = config
+        @io = io
+      end
+
+      def run
+        Dir.chdir(root) do
+          cmd = [
+            "grep",
+            "--color=always", # Highlight matches
+            "-E", # Use Extended Regular Expressions
+            "-H", # Print filename
+            "-n", # Print line number
+            "-r", # Read all files under each directory
+            "-s", # Suppress error messages
+            *pattern_args, # What to look for
+            *include_paths # Where to look
+          ]
+          Open3.popen2(*cmd) do |_in, out, _wait_thread|
+            out.each do |line|
+              report line
+            end
+          end
+        end
+      end
+
+      private
+
+      attr_reader :root, :config, :io
+
+      def patterns
+        config.dig("config", "patterns") || []
+      end
+
+      def include_paths
+        config.dig("include_paths") || ["."]
+      end
+
+      def message
+        config.dig("config", "output") || ""
+      end
+
+      def pattern_args
+        ["-e"].product(patterns).flatten
+      end
+
+      def all_matches(string)
+        [].tap do |matches|
+          match = MATCH_RE.match(string)
+          while match
+            matches << match
+            match = MATCH_RE.match(string, matches.last.end(0))
+          end
+        end
+      end
+
+      def report(match_line)
+        filename, line, string = match_line.split(":", 3)
+        filename = filename.gsub(MARKER_RE, "")
+        line = line.gsub(MARKER_RE, "").to_i
+        string = string.sub(/^#{MARKER_RE}/, "")
+
+        all_matches(string).each do |match|
+          real_begin = string[0..match.begin(0)].gsub(MARKER_RE, '').length
+          real_end = real_begin + match[:str].length
+
+          json = JSON.generate(
+            type: "issue",
+            check_name: "Pattern Match",
+            description: message,
+            categories: ["Style"], # FIXME
```

Comment:
I could imagine making this configurable by the user. This is a very flexible engine, and I can imagine a specific grep check fitting into any of the possible categories. What do you think about that idea?